### PR TITLE
sleep/hibernate-resume: pass hibernate location through efivar for resume without kernel cmdline

### PR DESCRIPTION
**Pull request by YHNdnzj**
This implements the efivar part of #27247.

### CODE REVIEW

File: src/sleep/sleep.c [hunk 1]
```diff
@@ -43,48 +44,69 @@
...
-        r = write_string_file("/sys/power/resume", resume_str, WRITE_STRING_FILE_DISABLE_BUFFER);
-        if (r < 0)
-                return log_debug_errno(r, "Failed to write partition device to /sys/power/resume for '%s': '%s': %m",
-                                       hibernate_location->swap->device, resume_str);
+        xsprintf(offset_str, "%" PRIu64, hibernate_location->offset);
 
-        log_debug("Wrote resume= value for %s to /sys/power/resume: %s", hibernate_location->swap->device, resume_str);
+#if ENABLE_EFI
+        if (is_efi_boot()) {
+                r = asprintf(&efi_hibernate_info, "%s%s%s", resume_str, need_offset ? "=" : "", strempty(offset_str));
```
Comment (poettering) [inline src/sleep/sleep.c]: major/minor numbers are not stable, you need to pick some other identifier, probably swap uuid or so. read from udev data.

File: src/sleep/sleep.c [hunk 2]
```diff
@@ -43,48 +44,69 @@
...
+
         xsprintf(resume_str, "%u:%u", major(hibernate_location->devno), minor(hibernate_location->devno));
-        r = write_string_file("/sys/power/resume", resume_str, WRITE_STRING_FILE_DISABLE_BUFFER);
-        if (r < 0)
-                return log_debug_errno(r, "Failed to write partition device to /sys/power/resume for '%s': '%s': %m",
-                                       hibernate_location->swap->device, resume_str);
+        xsprintf(offset_str, "%" PRIu64, hibernate_location->offset);
 
-        log_debug("Wrote resume= value for %s to /sys/power/resume: %s", hibernate_location->swap->device, resume_str);
+#if ENABLE_EFI
```
Comment (poettering) [inline src/sleep/sleep.c]: move this to some helper function, please

File: src/sleep/sleep.c [hunk 3]
```diff
@@ -43,48 +44,69 @@
...
+        if (is_efi_boot()) {
+                r = asprintf(&efi_hibernate_info, "%s%s%s", resume_str, need_offset ? "=" : "", strempty(offset_str));
+                if (r < 0)
+                        log_oom_debug();
```
Comment (poettering) [inline src/sleep/sleep.c]: whenever we hit OOM we should stop what we are doing, and give up, oom is not an error we can ignore.

File: src/sleep/sleep.c [hunk 4]
```diff
@@ -43,48 +44,69 @@
...
         char resume_str[DECIMAL_STR_MAX(unsigned) * 2 + STRLEN(":")];
-        int r;
+        char offset_str[DECIMAL_STR_MAX(uint64_t)];
+        bool need_offset;
+        int r = 0;
 
         assert(hibernate_location);
         assert(hibernate_location->swap);
 
+        need_offset = streq(hibernate_location->swap->type, "file");
```
Comment (poettering) [inline src/sleep/sleep.c]: why not simply specify the offset unconditionally? i.e. regardless what is referenced here. in case of partition it's going to be zero, but that#s totally fine, no?

File: src/hibernate-resume/hibernate-resume-generator.c [hunk 5]
```diff
@@ -41,6 +49,14 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
 
                 free_and_replace(arg_resume_device, s);
 
+        } else if (streq(key, "resume_offset")) {
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume-generator.c]: use proc_cmdline_key_streq() please

File: src/hibernate-resume/hibernate-resume-generator.c [hunk 6]
```diff
@@ -69,25 +85,84 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
...
+        if (r == -ENOENT)
+                return 0;
+        if (r < 0)
+                return log_error_errno(r, "Failed to get EFI variable HibernateInfo: %m");
+
+        info_split = strv_split(info, "=");
+        if (!info_split)
+                return log_oom();
+
+        r = parse_devnum(info_split[0], &dev);
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume-generator.c]: this will segfault if the string was empty, because then there is no element 0

File: src/hibernate-resume/hibernate-resume-generator.c [hunk 7]
```diff
@@ -5,25 +5,33 @@
...
+#include "path-util.h"
 #include "proc-cmdline.h"
 #include "special.h"
 #include "string-util.h"
+#include "strv.h"
 #include "unit-name.h"
 
 static const char *arg_dest = NULL;
 static char *arg_resume_device = NULL;
+static char *arg_resume_offset = NULL;
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume-generator.c]: please store this only in parsed form. it's a general rule we have.

File: src/hibernate-resume/hibernate-resume-generator.c [hunk 8]
```diff
@@ -69,25 +85,84 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
...
-        if (!lnk)
+        link = path_join(arg_dest, SPECIAL_SYSINIT_TARGET ".wants", service_unit);
+        if (!link)
                 return log_oom();
 
-        (void) mkdir_parents_label(lnk, 0755);
-        if (symlink(SYSTEM_DATA_UNIT_DIR "/systemd-hibernate-resume@.service", lnk) < 0)
-                return log_error_errno(errno, "Failed to create symlink %s: %m", lnk);
+        (void) mkdir_parents_label(link, 0755);
+        if (symlink(SYSTEM_DATA_UNIT_DIR "/systemd-hibernate-resume@.service", link) < 0)
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume-generator.c]: hmm, so the fact this has been a template unit has always been weird to me, because it is effectively a singleton: you cannot resume from multiple disks at the same time, hence there's no point in having multiple instances. The only reason it was implemented like this to allow easy parameterization via the block device name. But that's out of the window now because of the need to of a secon parameter. Hence I think we should just drop the instantiation. i.e. write out the full unit file, like we do for fsck and similar.

File: src/hibernate-resume/hibernate-resume.c [hunk 9]
```diff
@@ -29,17 +32,39 @@ int main(int argc, char *argv[]) {
...
 
         if (stat(device, &st) < 0) {
                 log_error_errno(errno, "Failed to stat '%s': %m", device);
                 return EXIT_FAILURE;
         }
 
         if (!S_ISBLK(st.st_mode)) {
-                log_error("Resume device '%s' is not a block device.", device);
+                log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Resume device '%s' is not a block device.", device);
                 return EXIT_FAILURE;
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume.c]: hmm, if you add proper error codes, pleas rework the whole main() function to use DEFINE_MAIN_FUNCTION() like we do in all modern code these days.

File: src/hibernate-resume/hibernate-resume.c [hunk 10]
```diff
@@ -29,17 +32,39 @@ int main(int argc, char *argv[]) {
...
         }
 
+        if (offset) {
+                if (access("/sys/power/resume_offset", W_OK) < 0) {
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume.c]: this has a TTOCTTOU feeling to it: drop the explicit initial access() check, and just write immediately to the sysattr. We'll get ENOENT from that much the same and can handle things exactly the same way, but with shorter code and without TTOCTTOU.

(i know the current sleep code does that too, but it really should be fixed there too)

File: src/hibernate-resume/hibernate-resume.c [hunk 11]
```diff
@@ -29,17 +32,39 @@ int main(int argc, char *argv[]) {
                 return EXIT_SUCCESS;
 
         device = argv[1];
+        offset = getenv("SYSTEMD_HIBERNATE_RESUME_OFFSET");
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume.c]: please parse this, and store it parsed.

**Review by poettering — CHANGES_REQUESTED** [review]

**Comment by YHNdnzj** [comment]
@poettering On a second thought, maybe we should suffix the EFI var name with `machine-id`? So that we can avoid the case that another Linux installation on the system hibernates and overwrites the previous one.

Is `machine-id` guaranteed to exist in initrd?

File: src/sleep/sleep.c [hunk 12]
```diff
@@ -44,49 +48,107 @@
...
+        _cleanup_free_ char *info = NULL;
+        char info_name[STRLEN("HibernateInfo-") + SD_ID128_STRING_MAX + STRLEN("-" EFI_VENDOR_SYSTEMD_STR)];
+        const char *uuid;
+        sd_id128_t machine_id = SD_ID128_NULL;
+
+        if (!is_efi_boot())
+                return 0;
+
+        errno = 0;
+        b = blkid_new_probe_from_filename(device);
```
Comment (poettering) [inline src/sleep/sleep.c]: hmm, use use blkid here, any specific reason? udev runs at this point already, so we can rely on the data udev already probed for us. that should be faster and simpler

File: src/sleep/sleep.c [hunk 13]
```diff
@@ -44,49 +48,107 @@
...
+        if (r < 0)
+                return log_full_errno(log_level,
+                                      errno_or_else(EIO),
+                                      "Failed to probe filesystem UUID of device '%s': %m", device);
+
+        r = asprintf(&info, "UUID=%s:%s", uuid, offset);
+        if (r < 0)
+                return log_oom();
+
+        r = sd_id128_get_machine(&machine_id);
```
Comment (poettering) [inline src/sleep/sleep.c]: as mentioned, the machine ID is not really a good bet here

File: src/sleep/sleep.c [hunk 14]
```diff
@@ -44,49 +48,107 @@
...
+        if (r < 0 && !ERRNO_IS_MACHINE_ID_UNSET(r))
+                return log_full_errno(log_level, r, "Failed to get machine-id: %m");
+
+        if (!sd_id128_is_null(machine_id))
+                xsprintf(info_name,
+                         "HibernateInfo-" SD_ID128_FORMAT_STR "-" EFI_VENDOR_SYSTEMD_STR,
+                         SD_ID128_FORMAT_VAL(machine_id));
+        else
+                strcpy(info_name, "HibernateInfo" "-" EFI_VENDOR_SYSTEMD_STR);
```
Comment (poettering) [inline src/sleep/sleep.c]: btw, I think we should be more precise here, and call this HibernateLocation or so, because that's really what it is about

File: src/sleep/sleep.c [hunk 15]
```diff
@@ -44,49 +48,107 @@
...
-        log_debug("Wrote resume_offset= value for %s to /sys/power/resume_offset: %s", hibernate_location->swap->device, offset_str);
+        if (r < 0) {
+                log_full_errno(r == -ENOENT ? LOG_DEBUG : LOG_ERR,
+                               r,
+                               "Failed to write swap file offset %s to /sys/power/resume_offset for device '%s': %m",
+                               offset_str, hibernate_location->swap->device);
+                return r == -ENOENT ? 0 : r;
+        }
+        log_debug("Wrote resume_offset=%s for device '%s' to /sys/power/resume_offset.",
+                  offset_str, hibernate_location->swap->device);
```
Comment (poettering) [inline src/sleep/sleep.c]: i appreciate that you reworked this, but can you do this in a separate commit?

File: src/hibernate-resume/hibernate-resume-generator.c [hunk 16]
```diff
@@ -130,6 +226,20 @@ static int run(const char *dest, const char *dest_early, const char *dest_late)
...
+        if (is_efi_boot()) {
+                r = parse_efi_hibernate_info();
+                if (r == -ENOMEM)
+                        return r;
+
+                /* Ensure that generic HibernateInfo will not survive more than one boot
+                 * to avoid spurious reuse. */
+                r = efi_set_variable(EFI_SYSTEMD_VARIABLE(HibernateInfo), NULL, 0);
+                if (r < 0)
+                        log_warning_errno(r, "Failed to clear EFI variable HibernateInfo, ignoring: %m");
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume-generator.c]: is this the right place to remove this? shouldn't this be done from the hibernation resume logic? i.e. in sleep.c?

File: src/hibernate-resume/hibernate-resume.c [hunk 17]
```diff
@@ -1,50 +1,58 @@
...
+
+        if (!S_ISBLK(st.st_mode))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "Resume device '%s' is not a block device.", arg_resume_device);
+
+        r = write_string_filef("/sys/power/resume_offset", WRITE_STRING_FILE_DISABLE_BUFFER, "%" PRIu64, arg_resume_offset);
+        if (r < 0)
+                return log_error_errno(r,
+                                       "Failed to write resume offset %" PRIu64 " to /sys/power/resume_offset for device '%s': %m",
+                                       arg_resume_offset, arg_resume_device);
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume.c]: hmm, this sysfs was added late iirc. maybe if we get ENOENT here and the offset is zero just ignore the failure (but still debug log)

**Comment by poettering** [comment]
btw, it would be even better if the cleanups (such as the conversion to using the main-func.h stuff) would be done in separate commits before the rest.

File: src/hibernate-resume/hibernate-resume-generator.c (hunk 16 above)
Comment (YHNdnzj) [reply src/hibernate-resume/hibernate-resume-generator.c]: I think we need to clear this if a normal boot happens, whether we would resume from it or not. Thus it can't be done in `systemd-sleep` or `systemd-hibernate-resume` 🤔

**Comment by poettering** [comment]
> Hmm, then I'd like to use `ID` as the varname identifier, and store the other 4 fields in json.
> 
> (I don't want to store everything into a single variable, because that might cause problem if one of the installed systemd versions has a buggy implementation.)

I'd really prefer if we wouldn't put variable data into variable names. The clean-up logic is a bit problematic then: you boot from a liveusb device once, and then have the efi far forever.

I mean, sure, some people want to dualboot two linux distros and hibernate switch between them, but also: meh, i don't care, sounds extremely specific, and they can always specify the hibernate device manually to avoid any confusion.

parsing json is relatively well tested in our codebase (i.e. comes with fuzzing and whatnot), hence I think it should not be risky. Also the implications of a parse failure would simply be that we revert to a regular boot instead of resuming hibernation. Which isn#t a total desaster, after all all file systems and so on should have synced before the hibernate safely...

File: src/sleep/sleep.c [hunk 18]
```diff
@@ -52,41 +52,31 @@ static int write_hibernate_location_info(const HibernateLocation *hibernate_loca
...
-        }
-
-        xsprintf(offset_str, "%" PRIu64, hibernate_location->offset);
         r = write_string_file("/sys/power/resume_offset", offset_str, WRITE_STRING_FILE_DISABLE_BUFFER);
-        if (r < 0)
-                return log_debug_errno(r, "Failed to write swap file offset to /sys/power/resume_offset for '%s': '%s': %m",
-                                       hibernate_location->swap->device, offset_str);
-
-        log_debug("Wrote resume_offset= value for %s to /sys/power/resume_offset: %s", hibernate_location->swap->device, offset_str);
+        if (r < 0) {
```
Comment (poettering) [inline src/sleep/sleep.c]: please keep the comment about kernel 4.17

File: src/sleep/sleep.c [hunk 19]
```diff
@@ -219,12 +209,14 @@ static int execute(
...
+
                 r = find_hibernate_location(&hibernate_location);
                 if (r < 0)
                         return log_error_errno(r, "Failed to find location to hibernate to: %m");
-                if (r == 0) { /* 0 means: no hibernation location was configured in the kernel so far, let's
-                               * do it ourselves then. > 0 means: kernel already had a configured hibernation
-                               * location which we shouldn't touch. */
+                resume_set = r > 0;
+
+                if (!resume_set) {
```
Comment (poettering) [inline src/sleep/sleep.c:265]: why not shorten this to:

```c
if (r > 0) { /* resume parameters have not been set in the kernel yet? */
```

File: src/sleep/sleep.c [hunk 20]
```diff
@@ -52,41 +52,31 @@ static int write_hibernate_location_info(const HibernateLocation *hibernate_loca
         assert(hibernate_location);
```
Comment (poettering) [inline src/sleep/sleep.c:68]: you could even make this wording stronger in the commit msg. it is inherently a *good* thing to override the resume offset in the kernel to zero, to undo possibly old half-written settings fully.

File: man/systemd-hibernate-resume-generator.xml [hunk 21]
```diff
@@ -55,6 +55,13 @@
         supported.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>resume_offset=</varname></term>
+
+        <listitem><para>Takes the offset of the swap space from the resume device.
+        Defaults to <literal>0</literal>.</para></listitem>
```
Comment (poettering) [inline man/systemd-hibernate-resume-generator.xml:63]: please mention the unit. (which is bytes, right?)

File: src/hibernate-resume/hibernate-resume.c [hunk 22]
```diff
@@ -1,58 +1,76 @@
...
+#include "main-func.h"
+#include "parse-util.h"
+#include "static-destruct.h"
+#include "string-util.h"
 
-int main(int argc, char *argv[]) {
+static char *arg_resume_device = NULL;
+static uint64_t arg_resume_offset = 0;
+
+STATIC_DESTRUCTOR_REGISTER(arg_resume_device, unsetp);
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume.c]: sounds unnecessary to unset this?

File: src/hibernate-resume/hibernate-resume.c [hunk 23]
```diff
@@ -1,58 +1,76 @@
...
 #include "fileio.h"
 #include "initrd-util.h"
 #include "log.h"
+#include "main-func.h"
+#include "parse-util.h"
+#include "static-destruct.h"
+#include "string-util.h"
 
-int main(int argc, char *argv[]) {
+static char *arg_resume_device = NULL;
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume.c]: if this points into argv[] you might as well make it `const char*`, given this should not be changed (at least not via this pointer I think)

File: src/hibernate-resume/hibernate-resume.c [hunk 24]
```diff
@@ -1,58 +1,76 @@
...
+                return 0;
 
-        device = argv[1];
+        arg_resume_device = argv[1];
 
-        if (stat(device, &st) < 0) {
-                log_error_errno(errno, "Failed to stat '%s': %m", device);
-                return EXIT_FAILURE;
-        }
+        r = safe_atou64(argv[2], &arg_resume_offset);
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume.c]: honestly, i'd make the second arg optional and just assume it#s zero.

This doesn't matter much, given that this is an internal command. but who knows, maybe someone glued together shell scripts that call the command directly. given it's easy we can just support the old command line, and just imply zero for the offset if only the device is specified.

File: src/hibernate-resume/hibernate-resume.c [hunk 25]
```diff
@@ -1,58 +1,76 @@
...
          * However, failed resume is a normal condition (may mean that there is
          * no hibernation image).
          */
 
-        log_info("Could not resume from '%s' (" DEVNUM_FORMAT_STR ").", device, DEVNUM_FORMAT_VAL(st.st_rdev));
-        return EXIT_SUCCESS;
+        log_info("Could not resume from device '%s' (" DEVNUM_FORMAT_STR ") offset %" PRIu64 ".",
+                 arg_resume_device,
+                 DEVNUM_FORMAT_VAL(st.st_rdev),
+                 arg_resume_offset);
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume.c]: maybe reword to "Unable to resume from hibernation device …, continuing boot process."

File: src/sleep/sleep.c [hunk 26]
```diff
@@ -52,41 +52,31 @@ static int write_hibernate_location_info(const HibernateLocation *hibernate_loca
...
 
+        if (!STR_IN_SET(hibernate_location->swap->type, "partition", "file"))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "Invalid swap type for hibernation: %s",
+                                       hibernate_location->swap->type);
+
         resume_str = FORMAT_DEVNUM(hibernate_location->devno);
+        xsprintf(offset_str, "%" PRIu64, hibernate_location->offset);
 
         r = write_string_file("/sys/power/resume", resume_str, WRITE_STRING_FILE_DISABLE_BUFFER);
```
Comment (poettering) [inline src/sleep/sleep.c]: hmm, how does this actually work again? isn't echoing the device info into it what triggers the resume? we don't actually want to do that here. how come this ever worked? /me is confused as he realizes he has no idea how this all actually works...

File: src/hibernate-resume/hibernate-resume.c [hunk 27]
```diff
@@ -1,58 +1,76 @@
...
-        }
+        if (stat(arg_resume_device, &st) < 0)
+                return log_error_errno(errno, "Failed to stat resume device '%s': %m", arg_resume_device);
 
-        r = write_string_file("/sys/power/resume", FORMAT_DEVNUM(st.st_rdev), WRITE_STRING_FILE_DISABLE_BUFFER);
+        if (!S_ISBLK(st.st_mode))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "Resume device '%s' is not a block device.", arg_resume_device);
+
+        r = write_string_filef("/sys/power/resume_offset", WRITE_STRING_FILE_DISABLE_BUFFER, "%" PRIu64, arg_resume_offset);
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume.c]: so here you write the offset first, the device second. in the sleep code the other way round. Wouldn't writing the offset first, and the device second be the safer choice always?

also, it feels as if we should have a common helper `int write_sleep_device(dev_t device, uint64_t offset)` or so?

File: src/sleep/sleep.c [hunk 28]
```diff
@@ -44,7 +50,101 @@
...
+        if (r < 0)
+                return log_full_errno(log_level, r, "Failed to create JSON field for filesystem UUID: %m");
+
+        r = json_variant_set_field(&v, "uuid", uuid_v);
+        if (r < 0)
+                return log_full_errno(log_level, r, "Failed to add uuid to JSON object: %m");
+
+        r = json_variant_set_field_unsigned(&v, "offset", hibernate_location->offset);
+        if (r < 0)
+                return log_full_errno(log_level, r, "Failed to add offset to JSON object: %m");
```
Comment (poettering) [inline src/sleep/sleep.c]: use json_build() instead, i.e.

```c
r = json_build(&v, JSON_BUILD_OJECT(
                JSON_BUILD_PAIR("uuid", JSON_BUILD_UUID(uuid)),
                JSON_BUILD_PAIR("offset", JSON_BUILD_UINSIGNED(hibernate->location->offset)),
                JSON_BUILD_PAIR_CONDITION(id, "osReleaseId", JSON_BUILD_STRING(image_id)),
                …
```

you get the idea

File: src/sleep/sleep.c [hunk 29]
```diff
@@ -44,7 +50,101 @@
...
+        r = parse_os_release(NULL,
+                             "ID", &id,
+                             "IMAGE_ID", &image_id,
+                             "VERSION_ID", &version_id,
+                             "IMAGE_VERSION", &image_version);
+        if (r < 0)
+                log_full_errno(log_level_ignore, r, "Failed to parse os-release, ignoring: %m");
+
+        if (id) {
+                r = json_variant_set_field_string(&v, "id", id);
```
Comment (poettering) [inline src/sleep/sleep.c]: that sounds overly generic. I'd just call these props osReleaseId, osReleaseImageId, and so on

File: src/hibernate-resume/hibernate-resume-generator.c [hunk 30]
```diff
@@ -29,6 +35,17 @@ STATIC_DESTRUCTOR_REGISTER(arg_resume_device, freep);
...
 STATIC_DESTRUCTOR_REGISTER(arg_root_options, freep);
 
+#if ENABLE_EFI
+typedef struct EFIHibernateLocation {
+        sd_id128_t uuid;
+        uint64_t offset;
+        char *id;
+        char *image_id;
+        char *version_id;
+        char *image_version;
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume-generator.c]: these four fields should be `const char*`, given you aprse them with json_dispatch_const_string

File: src/hibernate-resume/hibernate-resume-generator.c [hunk 31]
```diff
@@ -81,6 +98,90 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
...
+
+        r = json_dispatch(v, dispatch_table, NULL, JSON_LOG, &location);
+        if (r < 0)
+                return r;
+
+        xsprintf(node, "UUID=" SD_ID128_UUID_FORMAT_STR, SD_ID128_FORMAT_VAL(location.uuid));
+
+        device = fstab_node_to_udev_node(node);
+        if (!device)
+                return log_full_errno(log_level, SYNTHETIC_ERRNO(ENODEV), "Failed to get resume device through HibernateLocation.");
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume-generator.c]: just do:

```c
if (asprintf(&device, "/dev/disk/by-uuid/" SD_ID128_UUID_FORMAT_STR, SD_ID128_FORMAT_VAL(location.uuid)) < 0)
```

instead of first formatting this as fstab UUID= expressoin, and then converting it to /dev/disk/by-uuid/ syntax

File: src/hibernate-resume/hibernate-resume-generator.c [hunk 32]
```diff
@@ -81,6 +98,90 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
...
+                log_full_errno(log_level_ignore, r, "Failed to parse os-release, ignoring: %m");
+
+        if (!streq_ptr(id, location.id) ||
+            !streq_ptr(image_id, location.image_id) ||
+            !streq_ptr(version_id, location.version_id) ||
+            !streq_ptr(image_version, location.image_version)) {
+
+                log_notice("HibernateLocation system info mismatches with os-release, not resuming from it.");
+                return 0;
+        }
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume-generator.c:160]: please do this check before you start formatting the device, since conceptually it jus tmake sense to do the check first, and then use the data

File: src/sleep/sleep.c [hunk 33]
```diff
@@ -44,7 +50,101 @@
...
+                r = json_variant_set_field_string(&v, "versionId", version_id);
+                if (r < 0)
+                        log_full_errno(log_level_ignore, r, "Failed to add versionId to JSON object, ignoring: %m");
+        }
+
+        if (image_version) {
+                r = json_variant_set_field_string(&v, "imageVersion", image_version);
+                if (r < 0)
+                        log_full_errno(log_level_ignore, r, "Failed to add imageVersion to JSON object, ignoring: %m");
+        }
```
Comment (poettering) [inline src/sleep/sleep.c]: thinking about it we should probably include `uname -r` info here too and match it, i.e. kernel version.

i.e. uname()'s `release` field from struct utsname

File: src/hibernate-resume/hibernate-resume-generator.c [hunk 34]
```diff
@@ -81,6 +98,90 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
...
+
+        if (!arg_resume_device) {
+                arg_resume_device = TAKE_PTR(device);
+                arg_resume_offset = location.offset;
+        } else {
+                if (!streq(arg_resume_device, device))
+                        log_warning("resume=%s mismatches with HibernateLocation device '%s', proceeding anyway.",
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume-generator.c]: mismatches → doesn't match

File: src/hibernate-resume/hibernate-resume-generator.c [hunk 35]
```diff
@@ -81,6 +98,90 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
...
+                                    arg_resume_device, device);
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume-generator.c:171]: maybe say in the message explicitly, which of the two will be in effect

File: src/hibernate-resume/hibernate-resume-generator.c [hunk 36]
```diff
@@ -81,6 +98,90 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
...
+
+        _cleanup_(json_variant_unrefp) JsonVariant *v = NULL;
+        _cleanup_free_ char *location_str = NULL, *device = NULL, *id = NULL, *image_id = NULL,
+                       *version_id = NULL, *image_version = NULL;
+        char node[STRLEN("UUID=") + SD_ID128_UUID_STRING_MAX];
+        EFIHibernateLocation location = {};
+        int log_level, log_level_ignore;
+
+        log_level = arg_resume_device ? LOG_DEBUG : LOG_ERR;
+        log_level_ignore = arg_resume_device ? LOG_DEBUG : LOG_WARNING;
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume-generator.c]: i'd not bother with this, and just always log at LOG_ERR/LOG_WARNING level. If the data is in the efi var i see no shame in reading it, and hence log about it.

**Review by poettering — CHANGES_REQUESTED** [review]
I like this a lot, but a bunch of comments

File: src/sleep/sleep.c (hunk 26 above)
Comment (YHNdnzj) [reply src/sleep/sleep.c]: The kernel just simply ignores it (but preserves the value) if the written device doesn't contain a hibernation image already I think.

Comment (poettering) [reply src/sleep/sleep.c]: that's ugly is fuck though. is there no better way to configure this without triggering this behaviour?

i mean, are we even sure we destroy the hibernation image sig after resuming from it? does the kernel do that on its own?

Comment (YHNdnzj) [reply src/sleep/sleep.c]: > that's ugly is fuck though. is there no better way to configure this without triggering this behaviour?

Unfortunately I'm not aware of any...

> i mean, are we even sure we destroy the hibernation image sig after resuming from it? does the kernel do that on its own?

Yeah, it does: https://elixir.bootlin.com/linux/latest/source/kernel/power/swap.c#L1538

File: src/sleep/sleep.c (hunk 33 above)
Comment (YHNdnzj) [reply src/sleep/sleep.c]: Well, actually you can resume to a different kernel (I've tried that several times). The control is given back to the kernel from memory image after resume: https://elixir.bootlin.com/linux/latest/source/kernel/power/hibernate.c#L545

But one thing I'm not sure about is that whether this has any sorts of guarantee for backwards compatibility...

Comment (poettering) [reply src/sleep/sleep.c]: lets better be safe than sorry, and include the info, it's simple after all.

File: src/sleep/sleep.c (hunk 19 above)
Comment (YHNdnzj) [reply src/sleep/sleep.c:265]: This is used in later commits.

File: src/shared/sleep-util.c [hunk 37]
```diff
@@ -1028,6 +1028,45 @@ int read_fiemap(int fd, struct fiemap **ret) {
...
+        const char *devno_str;
+        int r;
+
+        devno_str = FORMAT_DEVNUM(devno);
+        xsprintf(offset_str, "%" PRIu64, offset);
+
+        if (!device) {
+                r = device_path_make_canonical(S_IFBLK, devno, &path);
+                if (r < 0)
+                        return log_oom();
```
Comment (poettering) [inline src/shared/sleep-util.c]: device_path_make_canonical() can actually return all kinds of errors, maybe use `log_error_errno(r, "Failed to format canonical device path for device %s: %m")`

File: src/shared/sleep-util.c [hunk 38]
```diff
@@ -1028,6 +1028,45 @@ int read_fiemap(int fd, struct fiemap **ret) {
...
+                device = path;
+        }
+
+        /* We write the offset first since it's safer. Note that this file is only available in 4.17+, so
+         * fail gracefully if it doesn't exist. */
+        r = write_string_file("/sys/power/resume_offset", offset_str, WRITE_STRING_FILE_DISABLE_BUFFER);
+        if (r == -ENOENT)
+                log_warning_errno(r, "/sys/power/resume_offset is unavailable, skipping writing swap file offset.");
```
Comment (poettering) [inline src/shared/sleep-util.c:1058]: hmm, i think we should make this safer by returning a fatal error if the sysfs attribute for this does not exist, and the offset is non-zero.

i.e.:

```c
if (r == -ENOENT) {
       if (offset != 0)
               return log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "Can't configure hibernation offset %" PRIu64, kernel does not support /sys/power/resume_offset.");
       log_warning_errno(r, "…);
}
…
```

you get the idea

File: man/systemd-hibernate-resume-generator.xml [hunk 39]
```diff
@@ -55,6 +55,13 @@
         supported.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>resume_offset=</varname></term>
+
+        <listitem><para>Takes the byte offset of the swap space from the resume device.
```
Comment (poettering) [inline man/systemd-hibernate-resume-generator.xml]: hmm, this is not correct. the unit is page size. see Documentation/admin-guide/kernel-parameters.txt in the kernel source tree.

File: man/systemd-hibernate-resume.service.xml [hunk 40]
```diff
@@ -3,41 +3,41 @@
...
+    <para><filename>systemd-hibernate-resume.service</filename>
+    initiates the resume from hibernation.</para>
 
     <para><filename>systemd-hibernate-resume</filename> only supports
     the in-kernel hibernation implementation, see
     <ulink url="https://docs.kernel.org/power/swsusp.html">Swap suspend</ulink>.
     Internally, it works by writing the major:minor of specified
-    device node to <filename>/sys/power/resume</filename>.</para>
+    device node to <filename>/sys/power/resume</filename>, along with
+    the offset (<filename>/sys/power/resume_offset</filename>) if supported.</para>
```
Comment (poettering) [inline man/systemd-hibernate-resume.service.xml]: maybe say "offset in memory pages" here too

File: src/hibernate-resume/hibernate-resume.c [hunk 41]
```diff
@@ -1,58 +1,55 @@
...
-#include "fileio.h"
 #include "initrd-util.h"
 #include "log.h"
+#include "main-func.h"
+#include "parse-util.h"
+#include "sleep-util.h"
 
-int main(int argc, char *argv[]) {
+static const char *arg_resume_device = NULL;
+static uint64_t arg_resume_offset = 0;
```
Comment (poettering) [inline src/hibernate-resume/hibernate-resume.c]: similar here, add a comment `/* in memory pages */`

the thing is simply that throughout our codebase we prefer to unify sizes/time values and so on onto a common base, i.e. bytes, µs and so on. This one here is an exception. (an OK one, given that apparently everything around the hibernation logic operates in pages, and this is directly visible because of the kernel cmdline option). But exceptions really deserve comments.

File: src/sleep/sleep.c [hunk 42]
```diff
@@ -45,7 +52,80 @@
...
+                             "IMAGE_ID", &image_id,
+                             "VERSION_ID", &version_id,
+                             "IMAGE_VERSION", &image_version);
+        if (r < 0)
+                log_full_errno(log_level_ignore, r, "Failed to parse os-release, ignoring: %m");
+
+        r = json_build(&v, JSON_BUILD_OBJECT(
+                               JSON_BUILD_PAIR_UUID("uuid", uuid),
+                               JSON_BUILD_PAIR_UNSIGNED("offset", hibernate_location->offset),
+                               JSON_BUILD_PAIR_STRING("kernelVersion", uts.release),
```
Comment (poettering) [inline src/sleep/sleep.c]: because you ignore the uname() errors above maybe change to:

```c
JSON_BUILD_PAIR_CONDITION(!isempy(uts.release), "kernelVersion", uts.release))
```

**Review by poettering — CHANGES_REQUESTED** [review]

**Comment by poettering** [comment]
lgtm! thanks!
